library/core/src/fmt/num.rs

```diff
@@ -643,25 +643,42 @@ fn fmt_u128(n: u128, is_nonnegative: bool, f: &mut fmt::Formatter<'_>) -> fmt::R
 }
 
 /// Partition of `n` into n > 1e19 and rem <= 1e19
+///
+/// Integer division algorithm is based on the following paper:
+///
+///   T. Granlund and P. Montgomery, “Division by Invariant Integers Using Multiplication”
+///   in Proc. of the SIGPLAN94 Conference on Programming Language Design and
+///   Implementation, 1994, pp. 61–72
+///
 fn udiv_1e19(n: u128) -> (u128, u64) {
     const DIV: u64 = 1e19 as u64;
-    let high = (n >> 64) as u64;
-    if high == 0 {
-        let low = n as u64;
-        return ((low / DIV) as u128, low % DIV);
-    }
-    let sr = 65 - high.leading_zeros();
-    let mut q = n << (128 - sr);
-    let mut r = n >> sr;
-    let mut carry = 0;
-
-    for _ in 0..sr {
-        r = (r << 1) | (q >> 127);
-        q = (q << 1) | carry as u128;
-
-        let s = (DIV as u128).wrapping_sub(r).wrapping_sub(1) as i128 >> 127;
-        carry = (s & 1) as u64;
-        r -= (DIV as u128) & s as u128;
-    }
-    ((q << 1) | carry as u128, r as u64)
+    const FACTOR: u128 = 156927543384667019095894735580191660403;
+
+    let quot = if n < 1 << 83 {
```

Comment:
Is there an actual value in this condition? I think on majority of the targets we support `u128_mulhi` will be faster than a 64-bit division anyway.

Reply:
Of course if you're using laptop/desktop PC `u128_mulhi` works very fast and thus this branch may be needless. However integer multiplication is still expensive operation on some modern processors. For example, on the Intel Knights Landing micro architecture (which is widely used for supercomputers and high-performance workstations), `MUL` instruction needs >7 cycles for generating result. Also it has only 1 specific port for multiplication.

https://agner.org/optimize/

In addition, As you can see in assembler output, this conditional branch generates only 2 instructions, and second call to this function always fallback to fast path because `u128::MAX / 10^19 < 2^83`. That means, if fast path exists, then these 2 calls results in 44 instructions in total, 58 otherwise.